aten/src/ATen/core/op_registration/adaption.h

```diff
@@ -57,5 +60,45 @@ check_tensor_options_and_extract_memory_format(
     return options.memory_format_opt();
   }
 }
+
+inline void check_or_update_common_device(optional<Device>& common_device, const at::Tensor& tensor, at::CheckedFrom c) {
+  if (!tensor.defined()) {
+    return;
+  }
+
+  if (tensor.is_cpu() && tensor.dim() == 0) {
+    // CPU 0-dimenstion tensor is not considered as common device
```

Comment:
nit: dimension

Also, the logic here only really applies for TensorIterator kernels, so I'm not sure it's appropriate to unconditionally opt into it.

Reply:
Originally I am thinking it means we need to rule all `TensorIterator` kernels out from automatic common device check/guard logic, which seems to be a lot of kernels?

But I do agree to start with some simple/clear semantic is the right approach. Perhaps we can start with very strict common device check, and opt out all `TensorIterator` kernels for now (today's device guard logic, strictly speaking, doesn't work for them anyway). And we can decide later whether we want to introduce an optional "allow_cpu_scalar" option in `native_functions.yaml` in the future.

Reply:
We'll need to make TensorIterator take care of guards itself, but this should be possible to do.